fog/sample-paykit/src/client.rs

```diff
@@ -362,6 +360,232 @@ impl Client {
+                Ok(Some((tx_out, proof))) => {
+                    if sci.tx_in.ring[idx] != tx_out {
+                        return Err(Error::SciGlobalIndexTxOutMismatch(
+                            sci.tx_out_global_indices[idx],
+                        ));
+                    }
+                    sci.tx_in.proofs.push(proof);
+                }
+            }
+        }
+
+        // Aggregate total required outlay
+        let mut outlay: HashMap<TokenId, u64> = Default::default();
+        outlay.insert(fee.token_id, fee.value);
+        for req_amount in sci.required_output_amounts.iter() {
+            *outlay
+                .entry(TokenId::from(req_amount.token_id))
+                .or_default() += req_amount.value;
+        }
+
+        // Compute the leftover from the signed input
+        let leftover = {
+            let token_id = TokenId::from(sci.pseudo_output_amount.token_id);
+            let sci_change = outlay.get(&token_id).cloned().unwrap_or(0);
+            if sci.pseudo_output_amount.value < sci_change {
+                return Err(Error::SciUnprofitable);
+            }
+            let value = sci.pseudo_output_amount.value - sci_change;
+            // We will use outlay to search our own wallet for required amounts, but
+            // we don't need to do that for this token id because it is supplied by the SCI
+            outlay.remove(&token_id);
+            Amount::new(value, token_id)
+        };
+
+        // Make fog resolver
+        let fog_uris = (&[&self.account_key.change_subaddress()])
+            .iter()
+            .filter_map(|addr| addr.fog_report_url())
+            .map(FogUri::from_str)
+            .collect::<core::result::Result<Vec<_>, _>>()?;
+        let fog_responses = self
+            .fog_report_conn
+            .fetch_fog_reports(fog_uris.into_iter())?;
+        let fog_resolver = FogResolver::new(fog_responses, &self.fog_verifier)?;
+
+        let block_version = BlockVersion::try_from(self.tx_data.get_latest_block_version())?;
+
+        let change_destination = ReservedDestination::from(&self.account_key);
+
+        // Make transaction builder
+        // TODO: Use RTH memos
+        let mut tx_builder = TransactionBuilder::new(
+            block_version,
+            fee,
+            fog_resolver,
+            EmptyMemoBuilder::default(),
+        )?;
+        tx_builder.set_tombstone_block(self.compute_tombstone_block()?);
+
+        // Add the presigned input
+        tx_builder.add_presigned_input(sci)?;
+
+        // Pay the leftover to ourselves
+        tx_builder.add_change_output(leftover, &change_destination, rng)?;
+
+        // Contribute our own inputs that may be required
+        for (token_id, value) in outlay {
+            // Arbitrarily choose 3 as the maximum number of inputs
+            // TODO: Should be based on fee scaling and fee choice
+            const TARGET_NUM_INPUTS: usize = 3;
```

Comment:
Any reason for this to not be `MAX_INPUTS - 1`? (1 being used by the SCI)

Reply:
so, i got this number from the `build_transaction` function in this client already (i'm not sure why it was 3 there)

but in this case, it does make sense for it to be less than `MAX_INPUTS - 1`, because there may be several token ids at work here, and so if they each produce `MAX_INPUTS-1` then we will go over `MAX_INPUTS`.

I'm inclined to just leave it for now